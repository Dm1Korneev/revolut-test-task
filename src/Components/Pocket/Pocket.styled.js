
import styled from 'styled-components';

export const Container = styled.div`
    background-color: rgb(21, 99, 244);
    display: flex;
    flex-direction: column;
    padding: 1rem;
`;
