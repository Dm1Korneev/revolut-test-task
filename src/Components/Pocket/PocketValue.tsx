import React from 'react';
import PropTypes from 'prop-types';

import { getCurrencySymbol } from 'Helpers';

const defaultProps = {
  value: 0,
};

const propTypes = {
  currency: PropTypes.string.isRequired,
  value: PropTypes.number,
};

function PocketValue(props) {
  const {
    currency, value,
  } = props;

  return (
    <span>
      {'You have '}
      <small>{getCurrencySymbol(currency)}</small>
      {value.toFixed(2)}
    </span>
  );
}

PocketValue.propTypes = propTypes;
PocketValue.defaultProps = defaultProps;

export default PocketValue;
