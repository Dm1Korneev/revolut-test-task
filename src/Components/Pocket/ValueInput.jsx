import React from 'react';
import PropTypes from 'prop-types';

import { floorPlus } from 'Helpers';

const defaultProps = {
  valuePrefix: '',
  value: 0,
};

const propTypes = {
  changeValue: PropTypes.func.isRequired,
  value: PropTypes.number,
  valuePrefix: PropTypes.string,
};

function ValueInput(props) {
  const {
    changeValue, value, valuePrefix,
  } = props;

  const valueChangeHandler = (event) => {
    const { valueAsNumber, value: valueFromInput } = event.target;
    if (valueFromInput === '') {
      changeValue(null);
    }
    if (Number.isNaN(valueAsNumber)) {
      return;
    }
    let newValue = valueAsNumber < 0 ? -valueAsNumber : valueAsNumber;
    newValue = floorPlus(newValue);
    changeValue(newValue);
  };

  const valueForInput = value === null ? '' : value;
  const valuePrefixString = value === 0 ? '' : valuePrefix;

  return (
    <div>
      {valuePrefixString}
      <input type="number" value={valueForInput} step="0.01" onChange={valueChangeHandler} />
    </div>
  );
}

ValueInput.propTypes = propTypes;
ValueInput.defaultProps = defaultProps;

export default ValueInput;
