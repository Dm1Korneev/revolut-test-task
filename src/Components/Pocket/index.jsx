import React from 'react';
import PropTypes from 'prop-types';

import CurrencySelect from 'Components/CurrencySelect';
import ValueInput from 'Components/ValueInput';
import PocketValue from './PocketValue';
import { Container, InfoLine, InputsLine } from './Pocket.styled';

const defaultProps = {
  value: null,
  valueHave: 0,
  children: null,
  valuePrefix: '',
};

const propTypes = {
  currency: PropTypes.string.isRequired,
  value: PropTypes.number,
  valueHave: PropTypes.number,
  currencies: PropTypes.arrayOf(PropTypes.string).isRequired,
  setPocketCurrency: PropTypes.func.isRequired,
  changeValue: PropTypes.func.isRequired,
  children: PropTypes.element,
  valuePrefix: PropTypes.string,
  inputValueLabel: PropTypes.string.isRequired,
  selectCurrencyLabel: PropTypes.string.isRequired,
};

function Pocket(props) {
  const {
    currency, value, valueHave, currencies, setPocketCurrency,
    children, changeValue, valuePrefix, selectCurrencyLabel, inputValueLabel,
  } = props;

  return (
    <Container>
      <InputsLine>
        <CurrencySelect
          currency={currency}
          currencies={currencies}
          setPocketCurrency={setPocketCurrency}
          label={selectCurrencyLabel}
        />
        <ValueInput
          value={value}
          valuePrefix={valuePrefix}
          changeValue={changeValue}
          label={inputValueLabel}
        />
      </InputsLine>
      <InfoLine>
        <PocketValue value={valueHave} currency={currency} />
        {children}
      </InfoLine>
    </Container>
  );
}

Pocket.propTypes = propTypes;
Pocket.defaultProps = defaultProps;

export default Pocket;
