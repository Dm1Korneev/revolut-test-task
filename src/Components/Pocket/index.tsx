import React, { FC, ReactNode } from 'react';

import CurrencySelect from 'Components/CurrencySelect';
import ValueInput from 'Components/ValueInput';
import PocketValue from './PocketValue';
import { Container, InfoLine, InputsLine } from './Pocket.styled';

export type Props = {
  currency: string;
  value?: number | null;
  valueHave?: number;
  currencies: Array<string>;
  setPocketCurrency: () => void;
  changeValue: () => void;
  valuePrefix?: string;
  inputValueLabel: string;
  selectCurrencyLabel: string;
  children?: ReactNode;
}

const Pocket: FC<Props> = (props): JSX.Element => {
  const {
    currency, value = null, valueHave = 0, currencies, setPocketCurrency,
    children, changeValue, valuePrefix, selectCurrencyLabel, inputValueLabel,
  } = props;

  return (
    <Container>
      <InputsLine>
        <CurrencySelect
          currency={currency}
          currencies={currencies}
          setPocketCurrency={setPocketCurrency}
          label={selectCurrencyLabel}
        />
        <ValueInput
          value={value}
          valuePrefix={valuePrefix}
          changeValue={changeValue}
          label={inputValueLabel}
        />
      </InputsLine>
      <InfoLine>
        <PocketValue value={valueHave} currency={currency} />
        {children}
      </InfoLine>
    </Container>
  );
};

export default Pocket;
