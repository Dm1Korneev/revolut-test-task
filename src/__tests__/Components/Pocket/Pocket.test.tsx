import React from 'react';
import { render } from '@testing-library/react';
import Pocket from 'Components/Pocket';

jest.mock('Components/CurrencySelect', () => mockComponent('CurrencySelect'));
jest.mock('Components/ValueInput', () => mockComponent('ValueInput'));
jest.mock('Components/Pocket/PocketValue', () => mockComponent('PocketValue'));

describe('Pocket component', () => {
  const props = {
    currency: 'USD',
    value: 100.87,
    valueHave: 2000.90,
    currencies: ['USD', 'EUR', 'GBP'],
    setPocketCurrency: jest.fn(),
    changeValue: jest.fn(),
    valuePrefix: '-',
    inputValueLabel: 'Test label input',
    selectCurrencyLabel: 'Test label select',
  };

  test('renders the component', () => {
    const { asFragment } = render(<Pocket {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
