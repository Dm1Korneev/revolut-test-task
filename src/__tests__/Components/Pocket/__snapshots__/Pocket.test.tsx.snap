// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pocket component renders the component 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa iqQZSj"
  >
    <div
      class="sc-bwzfXH eXyeCP"
    >
      <div
        originalcomponent="CurrencySelect"
        test-prop-currencies="[\\"USD\\",\\"EUR\\",\\"GBP\\"]"
        test-prop-currency="\\"USD\\""
        test-prop-label="\\"Test label select\\""
      />
      <div
        originalcomponent="ValueInput"
        test-prop-label="\\"Test label input\\""
        test-prop-value="100.87"
        test-prop-valueprefix="\\"-\\""
      />
    </div>
    <div
      class="sc-htpNat hKiEjg"
    >
      <div
        originalcomponent="PocketValue"
        test-prop-currency="\\"USD\\""
        test-prop-value="2000.9"
      />
    </div>
  </div>
</DocumentFragment>
`;
